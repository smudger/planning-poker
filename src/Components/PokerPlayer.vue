<script setup>
import { computed } from "vue";
import { minidenticon } from "minidenticons";
import { CheckIcon } from "@heroicons/vue/24/solid";

const props = defineProps({
  name: { type: String, default: undefined },
  estimate: { type: [String, Number], default: undefined },
  showEstimate: { type: Boolean, required: true },
});

const avatar = computed(
  () =>
    "data:image/svg+xml;utf8," + encodeURIComponent(minidenticon(props.name)),
);
</script>

<template>
  <div
    :class="[
      name ? 'bg-gray-800 border-gray-800' : 'border-dashed border-gray-300',
    ]"
    class="rounded-2xl border-2 justify-between flex w-72 items-center px-6 py-4 space-x-2"
  >
    <h3
      class="text-base font-semibold leading-7 tracking-tight text-white whitespace-nowrap truncate"
    >
      {{ name }}
    </h3>
    <div
      v-if="showEstimate && name"
      :class="[estimate ? 'bg-indigo-600' : '']"
      class="h-16 w-16 shrink-0 flex items-center justify-center rounded-full"
    >
      <span
        class="text-white font-mono font-semibold leading-7 tracking-tight text-xl"
      >
        {{ estimate }}
      </span>
    </div>
    <div v-else-if="name" class="relative">
      <img
        :class="[estimate ? 'ring-4 ring-indigo-600' : '']"
        :src="avatar"
        alt="avatar"
        class="h-16 w-16 shrink-0 rounded-full bg-white"
      />
      <span
        v-if="estimate"
        class="absolute right-0 top-0 block h-4 w-4 rounded-full bg-indigo-600 ring-2 ring-white text-white"
      >
        <CheckIcon />
      </span>
    </div>
    <div v-else class="h-16 w-16 shrink-0 rounded-full bg-gray-300" />
  </div>
</template>
