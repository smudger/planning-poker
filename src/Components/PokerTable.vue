<script setup>
import PokerPlayer from "@/Components/PokerPlayer.vue";

defineProps({
  players: { type: Array, required: true },
  showEstimates: { type: Boolean, required: true },
});
</script>

<template>
  <ul
    class="w-full flex flex-col justify-between items-center gap-4 sm:grid sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 2xl:grid-cols-6"
  >
    <li v-for="[playerId, player] in players" :key="playerId" class="mx-auto">
      <PokerPlayer
        :estimate="player.estimate"
        :name="player.name"
        :show-estimate="showEstimates"
      />
    </li>
  </ul>
</template>
