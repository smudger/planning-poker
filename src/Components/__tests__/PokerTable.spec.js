import { afterEach, describe, expect, it, vi } from "vitest";
import { shallowMount } from "@vue/test-utils";
import PokerTable from "@/Components/PokerTable.vue";
import PokerPlayer from "@/Components/PokerPlayer.vue";

vi.mock("minidenticons", () => {
  return { minidenticon: () => "image-url" };
});

describe("PokerTable", () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it("displays each player when show estimates is false", () => {
    const players = [
      [1, { name: "Constance", estimate: 5 }],
      [3, { name: "Leslie", estimate: undefined }],
      [4, { name: undefined, estimate: undefined }],
    ];

    const wrapper = shallowMount(PokerTable, {
      props: {
        players,
        showEstimates: false,
      },
    });

    const playerComponents = wrapper.findAllComponents(PokerPlayer);
    expect(playerComponents.length).toBe(3);
    expect(
      playerComponents.every((c) => c.props("showEstimate") === false),
    ).toBe(true);
    expect(playerComponents[0].props("name")).toBe("Constance");
    expect(playerComponents[0].props("estimate")).toBe(5);
    expect(playerComponents[0].props("isHighest")).toBe(false);
    expect(playerComponents[0].props("isLowest")).toBe(false);
    expect(playerComponents[1].props("name")).toBe("Leslie");
    expect(playerComponents[1].props("estimate")).toBeUndefined();
    expect(playerComponents[1].props("isHighest")).toBe(false);
    expect(playerComponents[1].props("isLowest")).toBe(false);
    expect(playerComponents[2].props("name")).toBeUndefined();
    expect(playerComponents[2].props("estimate")).toBeUndefined();
    expect(playerComponents[2].props("isHighest")).toBe(false);
    expect(playerComponents[2].props("isLowest")).toBe(false);
  });

  it("displays each player estimate when show estimates is true", () => {
    const players = [
      [1, { name: "Constance", estimate: 2 }],
      [2, { name: "Marie", estimate: 3 }],
      [3, { name: "Jane", estimate: 5 }],
      [4, { name: "Leslie", estimate: undefined }],
      [5, { name: undefined, estimate: undefined }],
    ];

    const wrapper = shallowMount(PokerTable, {
      props: {
        players,
        showEstimates: true,
      },
    });

    const playerComponents = wrapper.findAllComponents(PokerPlayer);
    expect(playerComponents.length).toBe(5);
    expect(
      playerComponents.every((c) => c.props("showEstimate") === true),
    ).toBe(true);
    expect(playerComponents[0].props("name")).toBe("Constance");
    expect(playerComponents[0].props("estimate")).toBe(2);
    expect(playerComponents[0].props("isHighest")).toBe(false);
    expect(playerComponents[0].props("isLowest")).toBe(true);
    expect(playerComponents[1].props("name")).toBe("Marie");
    expect(playerComponents[1].props("estimate")).toBe(3);
    expect(playerComponents[1].props("isHighest")).toBe(false);
    expect(playerComponents[1].props("isLowest")).toBe(false);
    expect(playerComponents[2].props("name")).toBe("Jane");
    expect(playerComponents[2].props("estimate")).toBe(5);
    expect(playerComponents[2].props("isHighest")).toBe(true);
    expect(playerComponents[2].props("isLowest")).toBe(false);
    expect(playerComponents[3].props("name")).toBe("Leslie");
    expect(playerComponents[3].props("estimate")).toBeUndefined();
    expect(playerComponents[3].props("isHighest")).toBe(false);
    expect(playerComponents[3].props("isLowest")).toBe(false);
    expect(playerComponents[4].props("name")).toBeUndefined();
    expect(playerComponents[4].props("estimate")).toBeUndefined();
    expect(playerComponents[4].props("isHighest")).toBe(false);
    expect(playerComponents[4].props("isLowest")).toBe(false);
  });

  it("does not calculate highest and lowest when there are no estimates", () => {
    const players = [
      [1, { name: "Constance", estimate: undefined }],
      [2, { name: "Marie", estimate: undefined }],
      [3, { name: "Jane", estimate: undefined }],
    ];

    const wrapper = shallowMount(PokerTable, {
      props: {
        players,
        showEstimates: true,
      },
    });

    const playerComponents = wrapper.findAllComponents(PokerPlayer);
    expect(playerComponents.length).toBe(3);
    expect(
      playerComponents.every((c) => c.props("showEstimate") === true),
    ).toBe(true);
    expect(playerComponents.every((c) => c.props("isHighest") === false)).toBe(
      true,
    );
    expect(playerComponents.every((c) => c.props("isLowest") === false)).toBe(
      true,
    );
  });

  it("does not calculate highest and lowest when there is only one unique estimate", () => {
    const players = [
      [1, { name: "Constance", estimate: 2 }],
      [2, { name: "Marie", estimate: 2 }],
      [3, { name: "Jane", estimate: undefined }],
    ];

    const wrapper = shallowMount(PokerTable, {
      props: {
        players,
        showEstimates: true,
      },
    });

    const playerComponents = wrapper.findAllComponents(PokerPlayer);
    expect(playerComponents.length).toBe(3);
    expect(
      playerComponents.every((c) => c.props("showEstimate") === true),
    ).toBe(true);
    expect(playerComponents.every((c) => c.props("isHighest") === false)).toBe(
      true,
    );
    expect(playerComponents.every((c) => c.props("isLowest") === false)).toBe(
      true,
    );
  });

  it("renders correctly", () => {
    const players = [
      [1, { name: "Constance", estimate: 5 }],
      [3, { name: "Leslie", estimate: undefined }],
      [4, { name: undefined, estimate: undefined }],
    ];

    const wrapper = shallowMount(PokerTable, {
      props: {
        players,
        showEstimates: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
