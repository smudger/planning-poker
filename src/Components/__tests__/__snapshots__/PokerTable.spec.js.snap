// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokerTable > renders correctly 1`] = `
<ul
  class="w-full flex flex-col justify-between items-center gap-4 sm:grid sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 2xl:grid-cols-6"
>
  
  <li
    class="mx-auto"
  >
    <poker-player-stub
      estimate="5"
      name="Constance"
      showestimate="true"
    />
  </li>
  <li
    class="mx-auto"
  >
    <poker-player-stub
      name="Leslie"
      showestimate="true"
    />
  </li>
  <li
    class="mx-auto"
  >
    <poker-player-stub
      showestimate="true"
    />
  </li>
  
</ul>
`;
